tests/transforms/test_compile.py

```diff
@@ -0,0 +1,406 @@
+    def test_multiple_passes(self, wires):
+        """Test that running multiple passes produces the correct results."""
+
+        qfunc = build_qfunc(wires)
+        dev = qml.device("default.qubit", wires=Wires(wires))
+
+        qnode = qml.QNode(qfunc, dev)
+
+        # Rotation merging will not occur at all until commuting gates are
+        # pushed through
+        pipeline = [merge_rotations, commute_controlled(direction="left"), cancel_inverses]
+
+        transformed_qfunc = compile(pipeline=pipeline, num_passes=2)(qfunc)
+        transformed_qnode = qml.QNode(transformed_qfunc, dev)
+
+        original_result = qnode(0.3, 0.4, 0.5)
+        transformed_result = transformed_qnode(0.3, 0.4, 0.5)
+        assert np.allclose(original_result, transformed_result)
+
+        names_expected = ["Hadamard", "CNOT", "RX", "PauliY", "CY"]
+        wires_expected = [
+            Wires(wires[0]),
+            Wires([wires[2], wires[1]]),
+            Wires(wires[0]),
+            Wires(wires[2]),
+            Wires([wires[1], wires[2]]),
+        ]
+
+        compare_operation_lists(transformed_qnode.qtape.operations, names_expected, wires_expected)
+
+    @pytest.mark.parametrize(("wires"), [["a", "b", "c"], [0, 1, 2], [3, 1, 2], [0, "a", 4]])
+    def test_decompose_into_basis_gates(self, wires):
+        """Test that running multiple passes produces the correct results."""
+
+        qfunc = build_qfunc(wires)
+        dev = qml.device("default.qubit", wires=Wires(wires))
+
+        qnode = qml.QNode(qfunc, dev)
+
+        pipeline = [commute_controlled(direction="left"), cancel_inverses, merge_rotations]
+
+        basis_set = ["CNOT", "RX", "RY", "RZ"]
+
+        transformed_qfunc = compile(pipeline=pipeline, basis_set=basis_set)(qfunc)
+        transformed_qnode = qml.QNode(transformed_qfunc, dev)
+
+        original_result = qnode(0.3, 0.4, 0.5)
+        transformed_result = transformed_qnode(0.3, 0.4, 0.5)
+        assert np.allclose(original_result, transformed_result)
+
+        names_expected = [
+            "RZ",
+            "RX",
+            "RZ",
+            "RZ",
+            "CNOT",
+            "RX",
+            "RZ",
+            "RY",
+            "RZ",
+            "RY",
+            "CNOT",
+            "RY",
+            "CNOT",
+        ]
+
+        wires_expected = [
+            Wires(wires[0]),
+            Wires(wires[0]),
+            Wires(wires[0]),
+            Wires(wires[2]),
+            Wires([wires[2], wires[1]]),
+            Wires(wires[0]),
+            Wires(wires[1]),
+            Wires(wires[2]),
+            Wires(wires[2]),
+            Wires(wires[2]),
+            Wires([wires[1], wires[2]]),
+            Wires(wires[2]),
+            Wires([wires[1], wires[2]]),
+        ]
+
+        compare_operation_lists(transformed_qnode.qtape.operations, names_expected, wires_expected)
+
+    def test_compile_template(self):
+        """Test that an empty pipeline returns the original function."""
```

Comment:
docstring doesn't seem to match the test!

Reply:
```suggestion
    def test_compile_template(self):
        """Test that functions with templates are correctly expanded and compiled."""
```